core/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java

```diff
@@ -129,31 +184,183 @@ public FieldSortBuilder setNestedPath(String nestedPath) {
         if (order != null) {
-            builder.field("order", order.toString());
+            builder.field(ORDER.getPreferredName(), order.toString());
         }
         if (missing != null) {
-            builder.field("missing", missing);
-        }
-        if (ignoreUnmapped != null) {
-            builder.field(SortParseElement.IGNORE_UNMAPPED.getPreferredName(), ignoreUnmapped);
+            if (missing instanceof BytesRef) {
+                builder.field(MISSING.getPreferredName(), ((BytesRef) missing).utf8ToString());
+            } else {
+                builder.field(MISSING.getPreferredName(), missing);
+            }
         }
         if (unmappedType != null) {
-            builder.field(SortParseElement.UNMAPPED_TYPE.getPreferredName(), unmappedType);
+            builder.field(UNMAPPED_TYPE.getPreferredName(), unmappedType);
         }
         if (sortMode != null) {
-            builder.field("mode", sortMode);
+            builder.field(SORT_MODE.getPreferredName(), sortMode);
         }
         if (nestedFilter != null) {
-            builder.field("nested_filter", nestedFilter, params);
+            builder.field(NESTED_FILTER.getPreferredName(), nestedFilter, params);
         }
         if (nestedPath != null) {
-            builder.field("nested_path", nestedPath);
+            builder.field(NESTED_PATH.getPreferredName(), nestedPath);
         }
         builder.endObject();
         return builder;
     }
+
+    @Override
+    public boolean equals(Object other) {
+        if (! (other instanceof FieldSortBuilder)) {
+            return false;
+        }
+        FieldSortBuilder builder = (FieldSortBuilder) other;
+        return (Objects.equals(this.fieldName, builder.fieldName) &&
+                Objects.equals(this.nestedFilter, builder.nestedFilter) &&
+                Objects.equals(this.nestedPath, builder.nestedPath) &&
+                Objects.equals(this.missing, builder.missing) &&
+                Objects.equals(this.order, builder.order) &&
+                Objects.equals(this.sortMode, builder.sortMode) &&
+                Objects.equals(this.unmappedType, builder.unmappedType));
+    }
+    
+    @Override
+    public int hashCode() {
+        return Objects.hash(this.fieldName, this.nestedFilter, this.nestedPath,
+                this.missing, this.order, this.sortMode, this.unmappedType);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(this.fieldName);
+        if (this.nestedFilter != null) {
+            out.writeBoolean(true);
+            out.writeQuery(this.nestedFilter);
+        } else {
+            out.writeBoolean(false);
+        }
+        out.writeOptionalString(this.nestedPath);
+        if (this.missing != null) {
+            out.writeBoolean(true);
+            out.writeGenericValue(this.missing);
```

Comment:
I think `writeGenericValue` handles null values, so you could omitt the surrounding check.